pkg/kubelet/kubelet.go

```diff
@@ -895,6 +888,64 @@ func NewMainKubelet(kubeCfg *kubeletconfiginternal.KubeletConfiguration,
 	return klet, nil
 }
 
+func finalizeTLSOptions(kubeCfg *kubeletconfiginternal.KubeletConfiguration, kubeDeps *Dependencies, klet *Kubelet, certDirectory string) error {
+	if kubeDeps.TLSOptions == nil {
+		return nil
+	}
+
+	// the ordering of the TLS config mutations below is important since both GetCertificate and GetConfigForClient can be set
+
+	if kubeCfg.ServerTLSBootstrap && utilfeature.DefaultFeatureGate.Enabled(features.RotateKubeletServerCertificate) {
+		var err error
+		klet.serverCertificateManager, err = kubeletcertificate.NewKubeletServerCertificateManager(klet.kubeClient, kubeCfg, klet.nodeName, klet.getLastObservedNodeAddresses, certDirectory)
+		if err != nil {
+			return fmt.Errorf("failed to initialize certificate manager: %v", err)
+		}
+		kubeDeps.TLSOptions.Config.GetCertificate = func(_ *tls.ClientHelloInfo) (*tls.Certificate, error) {
+			cert := klet.serverCertificateManager.Current()
+			if cert == nil {
+				return nil, fmt.Errorf("no serving certificate available for the kubelet")
+			}
+			return cert, nil
+		}
+	}
+
+	klet.clientCertificateCAContentProvider = kubeDeps.ClientCertificateCAContentProvider
+	if klet.clientCertificateCAContentProvider != nil {
+		// Populate PeerCertificates in requests, but don't reject connections without verified certificates
+		kubeDeps.TLSOptions.Config.ClientAuth = tls.RequestClientCert
+	}
+
+	if getCert, certFile, keyFile := kubeDeps.TLSOptions.Config.GetCertificate, kubeDeps.TLSOptions.CertFile, kubeDeps.TLSOptions.KeyFile; getCert == nil && len(certFile) > 0 && len(keyFile) > 0 {
+		var err error
+		klet.certKeyContentProvider, err = dynamiccertificates.NewDynamicServingContentFromFiles("kubelet-serving-cert", certFile, keyFile)
+		if err != nil {
+			return fmt.Errorf("failed to initialize dynamic serving cert: %v", err)
+		}
+	}
+
+	if klet.clientCertificateCAContentProvider != nil || klet.certKeyContentProvider != nil {
+		klet.dynamicCertificateController = dynamiccertificates.NewDynamicServingCertificateController(
+			kubeDeps.TLSOptions.Config,
+			klet.clientCertificateCAContentProvider,
+			klet.certKeyContentProvider,
+			nil,
+			record.NewEventRecorderAdapter(kubeDeps.Recorder),
+		)
+		kubeDeps.TLSOptions.Config.GetConfigForClient = klet.dynamicCertificateController.GetConfigForClient
+
+		// register if possible
+		if notifier, ok := klet.clientCertificateCAContentProvider.(dynamiccertificates.Notifier); ok {
+			notifier.AddListener(klet.dynamicCertificateController)
+		}
+		if notifier, ok := klet.certKeyContentProvider.(dynamiccertificates.Notifier); ok {
```

Comment:
need nil checks right?

Reply:
You can type check against `nil` without `panic`: https://goplay.space/#KJqD9ngavkI